OpenRA.Game/Orders/UnitOrderGenerator.cs

```diff
@@ -23,16 +23,30 @@ public IEnumerable<Order> Order(World world, CPos xy, MouseInput mi)
 				.Where(a => !world.FogObscures(a) && a.HasTrait<ITargetable>())
 				.WithHighestSelectionPriority();
 
+			var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
+					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
+					.WithHighestSelectionPriority();
+
 			Target target;
 			if (underCursor != null)
-				target = Target.FromActor(underCursor);
-			else
 			{
-				var frozen = world.ScreenMap.FrozenActorsAt(world.RenderPlayer, mi)
-					.Where(a => a.Info.Traits.Contains<ITargetableInfo>() && !a.Footprint.All(world.ShroudObscures))
-					.WithHighestSelectionPriority();
-				target = frozen != null ? Target.FromFrozenActor(frozen) : Target.FromCell(world, xy);
+				if (frozen != null)
+				{
+					int xDiffUC = underCursor.Bounds.X + underCursor.Bounds.Width / 2 - mi.Location.X;
+					int yDiffUC = underCursor.Bounds.Y + underCursor.Bounds.Height / 2 - mi.Location.Y;
+					int xDiffF = frozen.Bounds.X + frozen.Bounds.Width / 2 - mi.Location.X;
+					int yDiffF = frozen.Bounds.Y + frozen.Bounds.Height / 2 - mi.Location.Y;
+
+					if (xDiffUC * xDiffUC + yDiffUC * yDiffUC <= xDiffF * xDiffF + yDiffF * yDiffF)
+						target = Target.FromActor(underCursor);
+					else
+						target = Target.FromFrozenActor(frozen);
+				}
+				else
+					target = Target.FromActor(underCursor);
 			}
+			else
+				target = frozen != null ? Target.FromFrozenActor(frozen) : Target.FromCell(world, xy);
```

Comment:
`if { if { if { foo = a } else { foo = b } } else { foo = a } } else { if { foo = b } else { foo = c } } `

This logic can be simplified down to remove the redundancies and make the overall flow clearer.